### DISCUSSION TITLE
Proof of Concept: IPTV "xmltv.xml" EPG for yor AzuraCast Shows! Great with media centers like Jellyfin, KODI, …

### DISCUSSION THREAD

**Discussion by Moonbase59** [AzuraCast/AzuraCast#6665] [Ideas]
Some time ago, I suggested a feature [Add an IPTV "xmltv.xml" (or "xmltv.gz") endpoint, also usable for stations without a video stream](https://features.azuracast.com/suggestions/499503/add-an-iptv-xmltvxml-or-xmltvgz-endpoint-also-usable-for-stations-without-a-vide), mainly because I wanted to see what’s playing on my station in my _Jellyfin_ or _KODI_ (works as well with _xTeVe_, _Plex_, _Emby_, …).

I suggested AzuraCast might provide this (much like the `/station/{station_id}/schedule` API endpoint) dynamically, or as an `xmltv.xml` file.

To show that it’s feasible and can actually help making your station and shows more visible, I sat down today and wrote a quick-n-dirty Python script that polls the API and writes out an EPG for my station:

```xml
<?xml version='1.0' encoding='UTF-8'?>
<!DOCTYPE tv SYSTEM "xmltv.dtd">
<tv source-info-name="azuracast_xmltv" generator-info-name="azuracast_xmltv">
  <channel id="niteradio.de">
    <display-name>Nite Radio</display-name>
    <icon src="https://example.com/static/uploads/album_art.1678623041.png"/>
  </channel>
  <programme start="20231014180000 +0200" stop="20231015000000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">Heart Dance from London, UK</title>
  </programme>
  <programme start="20231014200000 +0200" stop="20231015000000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">Live: Martin 667</title>
    <credits>
      <presenter>Martin 667</presenter>
    </credits>
  </programme>
  <programme start="20231015000000 +0200" stop="20231015060000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">Nuit électronique</title>
  </programme>
  <programme start="20231015060000 +0200" stop="20231015120000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">Pop</title>
  </programme>
  <programme start="20231015060000 +0200" stop="20231015120000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">Schlager</title>
  </programme>
  <programme start="20231015120000 +0200" stop="20231015150000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">Easy Listening</title>
  </programme>
  <programme start="20231015150000 +0200" stop="20231015170000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">New Age</title>
  </programme>
  <programme start="20231015170000 +0200" stop="20231015200000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">Lounge</title>
  </programme>
  <programme start="20231015200000 +0200" stop="20231015220000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">House</title>
  </programme>
  <programme start="20231015220000 +0200" stop="20231016000000 +0200" channel="niteradio.de">
    <desc lang="en">Wir spielen die bessere Musik.</desc>
    <category lang="en">Music</category>
    <title lang="en">Goa, Psytrance</title>
  </programme>
</tv>
```

Looks like AzuraCast only delivers schedule data for today and the next day, even if I request 100 rows. Would be nicer if we could have a week, EPGs aren’t updated so often (I do it every 12 hours). Oh well, and would be great if we had a playlist _description_, which could be used in the EPG—for this test, I simply used a static text.

Time to let some screenshots speak:

**AzuraCast Schedule** (I added a streamer just to test out things)

![Auswahl_281](https://github.com/AzuraCast/AzuraCast/assets/3706922/49024b07-bc0f-42ab-8891-29af1e8ff02b)

**Program Overview for my Video Stream in Jellyfin**

![Jellyfin – Mozilla Firefox_004](https://github.com/AzuraCast/AzuraCast/assets/3706922/99e41bf4-d9fe-4ec7-8af4-bb639d715459)

(As you can see, I haven’t yet solved overlapping playlists: Sundays 6:00–12:00 _two_ playlists generate the program, namely _Pop_ and _Schlager_, and an XMLTV EPG expects only _one_ program running at the same time, so the last one is taken and "Schlager" shown.)

**EPG view in Jellyfin** (when I tested only with one channel, now they’re all in)

![Live-TV – Mozilla Firefox_002](https://github.com/AzuraCast/AzuraCast/assets/3706922/66d62b36-58cf-490b-8458-c8dc3fc4eeac)

**What’s playing display in Jellyfin**

![Jellyfin – Mozilla Firefox_005](https://github.com/AzuraCast/AzuraCast/assets/3706922/f9dc3ace-6863-4190-9932-8bf7db79e065)

**Family member playing my video stream 500 mi away, in their browser**

![Heart Dance from London, UK – Mozilla Firefox_001](https://github.com/AzuraCast/AzuraCast/assets/3706922/345f947c-bec8-41d7-81a1-e934b0045691)

This should show that it can be done, and is quite easy. The Python source is in my [azuracast_xmltv Gist](https://gist.github.com/Moonbase59/b76ab3b565dba65a9695b1f72d78d931). Be warned: Quick-n-dirty, many things hardcoded, just for a quick test.

Since pulling is always worse than getting fresh data from a built-in endpoint, it would be great if AzuraCast provided this.

**Please vote for this [on the feature request page](https://features.azuracast.com/suggestions/499503/add-an-iptv-xmltvxml-or-xmltvgz-endpoint-also-usable-for-stations-without-a-vide) if you like what you see!**

**Comment by Moonbase59** [comment]
Almost forgot: For IPTV type clients, a **M3U playlist file** is also required. Generating this from the visible streams is a piece of cake, though. Example:

```m3u
#EXTM3U
#EXTINF:-1 tvg-name="Nite Radio (MP3)" tvg-id="niteradio.de" group-title="Nite Radio" tvg-logo="https://example.com/static/uploads/album_art.1678623041.png",Nite Radio (MP3)
https://example.com/listen/niteradio/radio.mp3
#EXTINF:-1 tvg-name="Nite Radio (AAC)" tvg-id="niteradio.de" group-title="Nite Radio" tvg-logo="https://example.com/static/uploads/album_art.1678623041.png",Nite Radio (AAC)
https://example.com/listen/niteradio/radio.aac
#EXTINF:-1 tvg-name="Nite Radio Video-Stream" tvg-id="niteradio.de" group-title="Nite Radio" tvg-logo="https://example.com/static/uploads/album_art.1678623041.png",Nite Radio Video-Stream
https://example.com/listen/niteradio/video.m2t
#EXTINF:-1 tvg-name="Nite Radio Testbild" tvg-id="niteradio.de" group-title="Nite Radio" tvg-logo="https://example.com/static/uploads/album_art.1678623041.png",Nite Radio Testbild
https://example.com/listen/niteradio/testbild.m2t
#EXTINF:-1 tvg-name="Nite Radio Cam1" tvg-id="niteradio.de" group-title="Nite Radio" tvg-logo="https://example.com/static/uploads/album_art.1678623041.png",Nite Radio Cam1
https://cam1.example.com/stream
```

**Comment by Moonbase59** [comment]
**A more complete, much improved version will be on Github soon!**

It will also handle auto-generating the M3U (tuner) playlist file(s) for each station and can be used to automatically update the EPG via a cron job.

**Comment by BusterNeece** [comment]
@Moonbase59 The `schedule` endpoint does support a `start` and `end` parameter that you can submit in YYYY-MM-DD format.

**Reply by Moonbase59** [reply]
How cool is that! Will save me some looping… Wish there would be more info on all these wonderful API secrets ;-) Will try tomorrow.

**Reply by Moonbase59** [reply]
Couldn’t keep my fingers off it and did a few `curl`s. Seems there are some bugs/oddities in that still:

- `curl -sX 'GET' 'https://example.com/api/station/1/schedule?start=2023-10-22&end=2023-10-22' | jq` has some other days in it (07..23)
- `curl -sX 'GET' 'https://radio.niteradio.net/api/station/1/schedule?start=2023-10-22T00:00:01%2B02:00&end=2023-10-22T10:00:00%2B02:00' | jq` gives entries from days 21..23 but far outside the hour range
- and I had one (I _think_ it was 2023-10-22 to 2023-10-23) that gave me one entry from the 15th

Will write a quick sort to check tomorrow, it’s 2:30 a.m. and too late to sort all results in the brain…

Btw, it did get the switch from DST to ST correct (on 29th Oct here), switched from `+02:00` to `+01:00` alright. Time to watch our old schedule bug again. :-)

**Reply by Moonbase59** [reply]
Created an issue https://github.com/AzuraCast/AzuraCast/issues/6685 for this, to better keep focused.
